packages/blade/src/utils/useBreakpoint/useBreakpoint.ts

```diff
@@ -0,0 +1,142 @@
+import { useEffect, useState, useCallback, useMemo } from 'react';
+import type { Breakpoints } from '../../tokens/global';
+
+type DeviceType = 'desktop' | 'mobile';
+type Breakpoint = keyof Breakpoints | undefined;
+
+type BreakpointAndDevice = {
+  matchedBreakpoint: Breakpoint;
+  matchedDeviceType: DeviceType;
+};
+
+const useBreakpoint = ({ breakpoints }: { breakpoints: Breakpoints }): BreakpointAndDevice => {
+  const supportsMatchMedia =
+    typeof document !== 'undefined' && typeof window?.matchMedia === 'function';
+
+  const breakpointsTokenAndQueryCollection = useMemo(
+    () =>
+      (supportsMatchMedia
+        ? Object.entries(breakpoints).map(
+            ([breakpointTokenName, breakpointSize], index, breakpointsArray) => {
+              let mediaQuery = '';
+
+              if (breakpointTokenName === 'max') {
+                mediaQuery = `screen and (min-width: ${breakpointSize}px)`;
+              } else if (breakpointsArray[index - 1]) {
+                mediaQuery = `screen and (min-width: ${
+                  breakpointsArray[index - 1][1] + 1
+                }px) and (max-width: ${breakpointSize}px)`;
+              } else {
+                mediaQuery = `screen and (max-width: ${breakpointSize}px)`;
+              }
+
+              return [breakpointTokenName, breakpointSize, mediaQuery];
```

Comment:
Will probably be a bit better if we make the structure of this into an object? 

`{tokenName, size, query}`

Arrays are evil, need to keep track of indices.